simulator/recorder/recorder.go

```diff
@@ -138,40 +138,61 @@ func (s *Service) recordEvent(obj interface{}, e Event) {
 }
 
 func (s *Service) record(ctx context.Context) {
+	records := make([]Record, 0, s.recordBatchCapacity)
+	flushBuffer := func() {
+		if err := appendToFile(s.path, records); err != nil {
+			klog.Errorf("failed to append record to file: %v", err)
+		}
+		records = make([]Record, 0, s.recordBatchCapacity)
+	}
+
 	for {
 		select {
 		case r := <-s.recordCh:
-			if err := appendToFile(s.path, r); err != nil {
-				klog.Errorf("failed to append record to file: %v", err)
+			records = append(records, r)
+			if len(records) >= s.recordBatchCapacity {
+				flushBuffer()
 			}
 
 		case <-ctx.Done():
 			// flush the buffer
-			for len(s.recordCh) > 0 {
-				r := <-s.recordCh
-				if err := appendToFile(s.path, r); err != nil {
-					klog.Errorf("failed to append record to file: %v", err)
-				}
+			for r := range s.recordCh {
```

Comment:
I guess this line blocks until `s.recordCh` is closed.
And currently `s.recordCh` is not closed anywhere, so indefinitely blocked?

Reply:
Right. I changed the loop condition to `len(s.recordCh) > 0`.

Reply:
I guess this loop still does not break indefinitely because `len(records)` grows inside the loop.
Handling channels is error prone, so can we just drop the elements in `s.recordCh` and return when the context is cancelled?
We can implement graceful channel handling in another pull request.

Reply:
Sorry this is a just mistake. I was trying to change to `len(s.recordCh)`. How about it?

Reply:
It will work now!